docs/architecture-decision-records/20230123-pipeline-config-files.md

```diff
@@ -0,0 +1,60 @@
+# Pipeline Config Files
+
+## Summary
+
+This document describes the behavior of provisioning pipelines using pipeline
+config files.
+
+## Context
+
+We want to provide a way to provision pipelines using config files in addition
+to the API. This unlocks the ability to treat pipelines like other
+configurations and store them in a repository, version them, do code reviews on
+changes, deploy, test etc. Doing the same thing through the API is error-prone
+and tedious.
+
+## Decision
+
+Pipeline config files are loaded at the start of Conduit. The end result is a
+provisioned pipeline that matches whatever is defined in the config files (as
+long as the config files are valid).
+
+In a pipeline config files no changes are forbidden, config files are treated as
+the source of truth. This means that if you can change a config file between
+Conduit runs in any way you want, knowing that after the start Conduit will load
+the structure in the new config file. The worst thing that can happen is that
+you make an "invalid change" (e.g. you change the source connector plugin) and
+the state of that entity gets erased. That shouldn't be unexpected though,
+because you changed the plugin anyway, so you can't really expect to start from
+where you left off.
+
+Entities created through config files are stored in the Conduit store so that
+state can be preserved between runs.
+
+If a pipeline config file is loaded and we detect that Conduit already contains
+that pipeline in the store, the existing pipeline is updated to match whatever
+is in the config file.
```

Comment:
> If a pipeline config file is loaded and we detect that Conduit already contains
that pipeline in the store

Is the following correct?

1. If the already existing pipeline has been created through the API, then provisioning fails.
2. If the already existing pipeline has been created through a pipeline config file, then the pipeline is updated.

Reply:
1. Good questions, I looked at the code and it seems like we don't check how the pipeline was provisioned before updating it, so we would overwrite it even if it was previously created through the API. Not really the behavior I'd expect 🤔
2. Yes, this case is true (since it apparently doesn't matter how the pipeline was provisioned in the first place).

Reply:
1. same here, I wouldn't expect that.

Reply:
I created a bug report for point 1: https://github.com/ConduitIO/conduit/issues/837

Reply:
So regarding #1, I created a pipeline through the API and then restarted Conduit with a pipeline config file that contains a pipeline with the same name. I got "pipeline name already exists", no updates were done.